### DISCUSSION TITLE
Soundbar unavailable (Yamaha SR-X50A)

### DISCUSSION THREAD

**Discussion by martenos** [osk2/yamaha-soundbar#17] [General]
Hi,

I add this integration and set configuration, unfortunately my soundbar cannot be contacted as seems to be unavailable.
Here is a part of the log:

2024-11-29 13:41:17.470 ERROR (MainThread) [custom_components.yamaha_soundbar.media_player] For: My Sound Bar (192.168.0.85) async get failed, response code: 404 Full message: <ClientResponse(https://192.168.0.85/httpapi.asp?command=getPlayerStatus) [404 Not Found]>
<CIMultiDictProxy('Date': 'Fri, 29 Nov 2024 12:41:12 GMT', 'Server': 'Boa/0.94.13', 'Connection': 'close', 'Content-Type': 'text/html; charset=ISO-8859-1')>

can you help, please?

**Comment by martenos** [comment]
is this still managed project?

**Comment by osk2** [comment]
This component uses the `getPlayerStatus` API to check the soundbar's status. It's implemented on my YAS-209, but I can't test it further without another Yamaha soundbar.

I can apply a fix if you find out which API your device uses to read the status.

**Comment by LucaTino** [comment]
@martenos did you find a way? I'm interested too since I own a SR-X40A.
Thanks

**Reply by martenos** [reply]
no, unfortunately I haven't found which API it is using...

**Comment by zztopgtb** [comment]
Hi, Im getting this same error for the SR-X50A sound bar. How can I find out the API you need to get this working?

Thanks

**Reply by osk2** [reply]
As @ashiaraiyashiki mentioned [here](https://github.com/osk2/yamaha-soundbar/issues/20#issuecomment-2660869081). It seems SR-X50A uses different APIs. I have no plan to support it, but if you want to try, you can find the API endpoint by intercepting HTTP traffic between app and the soundbar. You can use tools like mitmproxy to do this.

**Reply by conioannou59** [reply]
I too am having this problem with my SR-X50A. So also looking for a solution.
My son works for Yamaha and may be able to get hold of documenttion. If someone can let me know specifically what would be needed I could see if it he can get hold of it.

**Comment by LaTrappe** [comment]
I got my SR-X50A since yesterday and I fiddled around with intercepting the HTTP traffic from the Soundbar Controller app to the soundbar. I was able to obtain a few working methods, which allow me to obtain some info and change a few settings:

### GetInfo
`https://{ip}/lp.asp?dst=linkplay.system&iface=linkplay.system&method=getInfo`

As maintainer described this yield some basic info of the soundbar. Apart from names, MAC address it does not provide that much usefull (i think).

### Get (and set) SoundSettings

To obtain the current settings

`https://{ip}/lp.asp?dst=linkplay.adk&iface=linkplay.adk&obj=/qti/iotsys/soundSetting&method=getSoundSetting&payload={}`

Provides the settings as found on the remote. 
```
{
  "status": "ok",
  "soundSetting": {
    "soundProgram": "Movie",
    "soundPath": "TV",
    "powerSaving": true,
    "clearVoice": true,
    "bassExtension": true,
    "nightMode": false
  }
}
```

The payload would indicate there's room to change settings. I have tried to set "clearVoice" to true but that didn't work. However, it does work when you change the method to the desired setting, it can be changed. 

To select the soundProgram, use: 

`https://{IP}/lp.asp?dst=linkplay.adk&iface=linkplay.adk&obj=/qti/iotsys/soundSetting&method=setSoundProgram&payload={"soundProgram":"STANDARD"}`

The following options are available (mind the capitals): MOVIE, STANDARD, GAME, STEREO, ALLMODE (The latter only available when surround speakers are connected). 

Same applies to setSoundPath which has: HMI, BT and TV

PowerSaving, NightMode, BassExtension and ClearVoice can be enabled / disabled in the following way:

`https://192.168.2.28/lp.asp?dst=linkplay.adk&iface=linkplay.adk&obj=/qti/iotsys/soundSetting&method=setClearVoice&payload={"clearVoice":true}`

### Change volumes of the wireless speakers

To obtain the current state:

`https://{ip}/lp.asp?dst=linkplay.adk&iface=linkplay.adk&obj=/qti/iotsys/allplay&method=getAllplaySetting`

This provides a list with the subwoofer and surround speaker properties. Here's a snipped of only my left rear speaker (the rest is there). 
 
```

      "surroundL": {
        "paired": true,
        "connected": true,
        "power_info": true,
        "model_name": "WS-X1A",
        "volume": 10,
        "fw_version": "01.00",
        "wifi_signal": 100
      }
```

To change the volume of the surroundL speaker use: 

`https://{ip}/lp.asp?dst=linkplay.adk&iface=linkplay.adk&obj=/qti/iotsys/allplay&method=setSurroundLVolume&payload={"surroundLVolume":"-5"}`

This also applies to setSubwooferVolume and setSurroundRVolume

### UPNP

The soundbar uses a standard UPNP protocol (I think its standard) for:

- Progress bar etc
- Desired volume (main volume of the speaker)
- More ?

It runs on:

`http://{IP}:49152/upnp/control/rendercontrol1`

Volume for instance can be changed using this body

```
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<s:Envelope s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/" xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
  <s:Body>
    <u:SetVolume xmlns:u="urn:schemas-upnp-org:service:RenderingControl:1">
      <InstanceID>0</InstanceID>
      <Channel>Single</Channel>
      <DesiredVolume>60</DesiredVolume>
    </u:SetVolume>
  </s:Body>
</s:Envelope> 
```

### Other commands

Reboot seems to work (funny cuz its not in the app). Also tried 'sleep', 'power' and 'poweron' etc but those didn't work. 

`https://{ip}/lp.asp?dst=linkplay.system&iface=linkplay.system&method=reboot`

List of streaming services (the results however do not match whats shown in my app - region thingy?)

`https://{ip}/lp.asp?dst=linkplay.streamservice&iface=linkplay.streamservice&method=streamServicesCapability`

This one appears on boot of the app, no clue what it exactly does suspect its an Alexa thing.

`https://192.168.2.28/lp.asp?dst=linkplay.AVS&iface=linkplay.common&method=propertyGet&payload={"name":"avsCBL"}`

The result is this:

```
{
  "state": 3,
  "msg": "REFRESHING_TOKEN"
}
```
